urllib3/response.py

```diff
@@ -216,14 +216,14 @@ def read(self, amt=None, decode_content=None, cache_content=False):
             self._fp_bytes_read += len(data)
 
             try:
-                if decode_content and self._decoder:
+                if data and decode_content and self._decoder:
                     data = self._decoder.decompress(data)
             except (IOError, zlib.error) as e:
                 raise DecodeError(
                     "Received response with content-encoding: %s, but "
                     "failed to decode it." % content_encoding, e)
 
-            if flush_decoder and decode_content and self._decoder:
+            if data and flush_decoder and decode_content and self._decoder:
```

Comment:
I think I understand your concern here better now. Basically, "What if we have data that's been decoded that hasn't been flushed yet?"

I think that's a valid usecase and that your problems were mostly arising from the block above, right? If you remove this check, do your tests fail?